memory/src/main/java/io/airlift/memory/jetty/ConcurrentRetainableBufferPool.java

```diff
@@ -0,0 +1,343 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.airlift.memory.jetty;
+
+import io.airlift.units.DataSize;
+import org.eclipse.jetty.io.ByteBufferPool;
+import org.eclipse.jetty.io.RetainableByteBuffer;
+
+import java.lang.foreign.Arena;
+import java.lang.foreign.MemorySegment;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static io.airlift.units.DataSize.Unit.MEGABYTE;
+
+public class ConcurrentRetainableBufferPool
+        implements ByteBufferPool
+{
+    private static final int MIN_POOL_SIZE_SHIFT = 7;
+    private static final int[] poolSizeShiftToSize;
+
+    private final int numBuckets;
+    private final int checkMaxMemoryPoint;
+    private final ArenaBucket[] heapBuckets;
+    private final ArenaBucket[] offHeapBuckets;
+    private final AtomicBoolean evictor = new AtomicBoolean(false); // left from the original jetty code
+    private long maxHeapMemory;
+    private long maxOffHeapMemory;
+    private int checkCount;
+
+    static {
+        poolSizeShiftToSize = new int[Integer.SIZE - MIN_POOL_SIZE_SHIFT];
+        for (int i = 0; i < poolSizeShiftToSize.length; i++) {
+            poolSizeShiftToSize[i] = 1 << i + MIN_POOL_SIZE_SHIFT;
+        }
+    }
+
+    public ConcurrentRetainableBufferPool(long maxHeapMemory, long maxOffHeapMemory)
+    {
+        this.numBuckets = Runtime.getRuntime().availableProcessors() * 4; // factor of number of tasks
+        this.checkMaxMemoryPoint = numBuckets * 100;
+        // note that both max numbers are adaptive to avoid getting too many evictions. in case max buffer is larger, max memory is updated as well
+        this.maxHeapMemory = (maxHeapMemory > 0) ? maxHeapMemory : DataSize.of(1, MEGABYTE).toBytes();
+        this.maxOffHeapMemory = (maxOffHeapMemory > 0) ? maxOffHeapMemory : DataSize.of(1, MEGABYTE).toBytes();
+        this.checkCount = 0;
+
+        heapBuckets = new ArenaBucket[numBuckets];
+        offHeapBuckets = new ArenaBucket[numBuckets];
+        for (int bucketId = 0; bucketId < numBuckets; bucketId++) {
+            heapBuckets[bucketId] = new ArenaBucket(false, bucketId);
+            offHeapBuckets[bucketId] = new ArenaBucket(true, bucketId);
+        }
+    }
+
+    @Override
+    public RetainableByteBuffer acquire(int size, boolean isOffHeap)
+    {
+        ArenaBucket bucket = getBucketArray(isOffHeap)[(int) (Thread.currentThread().threadId() % numBuckets)];
+        return bucket.alloc(size);
+    }
+
+    private ArenaBucket[] getBucketArray(boolean isOffHeap)
+    {
+        return isOffHeap ? offHeapBuckets : heapBuckets;
```

Comment:
instead of this weird accessor just do

```
ArenaBucket bucket = isOffHeap ? offHeapBuckets[floorMod(Thread.currentThread().threadId(), numBuckets)] : heapBuckets[floorMod(Thread.currentThread().threadId(), numBuckets)];
```